File: backend/apps/industries/schemas.py
from datetime import datetime
from typing import Any
from uuid import UUID

from ninja import Schema
from pydantic import Field


class IndustryServiceInputSchema(Schema):
    service_id: UUID
    description: str = ""
    sort_order: int = Field(default=0, ge=0)
    is_featured: bool = False


class IndustryFaqInputSchema(Schema):
    question: str = Field(min_length=1, max_length=300)
    answer: str = Field(min_length=1)
    sort_order: int = Field(default=0, ge=0)


class IndustrySeoInputSchema(Schema):
    meta_title: str = Field(default="", max_length=70)
    meta_description: str = Field(
        default="",
        max_length=170,
    )
    canonical_url: str = ""
    robots_index: bool = True
    robots_follow: bool = True
    open_graph_title: str = Field(
        default="",
        max_length=100,
    )
    open_graph_description: str = Field(
        default="",
        max_length=200,
    )
    open_graph_image_id: UUID | None = None
    twitter_title: str = Field(
        default="",
        max_length=100,
    )
    twitter_description: str = Field(
        default="",
        max_length=200,
    )
    structured_data: dict[str, Any] = Field(
        default_factory=dict,
    )


class IndustryCreateSchema(Schema):
    name: str = Field(min_length=1, max_length=250)
    slug: str = Field(min_length=1, max_length=250)
    short_description: str = Field(
        default="",
        max_length=350,
    )
    description: dict[str, Any] = Field(
        default_factory=dict,
    )
    hero_title: str = Field(default="", max_length=250)
    hero_description: str = ""
    hero_image_id: UUID | None = None
    icon: str = Field(default="", max_length=100)
    status: str = "draft"
    is_featured: bool = False
    is_active: bool = True
    sort_order: int = Field(default=0, ge=0)
    challenges: list[Any] = Field(default_factory=list)
    solutions: list[Any] = Field(default_factory=list)
    benefits: list[Any] = Field(default_factory=list)
    cta_title: str = Field(default="", max_length=200)
    cta_text: str = Field(default="", max_length=300)
    cta_label: str = Field(default="", max_length=100)
    cta_url: str = Field(default="", max_length=500)
    services: list[IndustryServiceInputSchema] = Field(
        default_factory=list,
    )
    faqs: list[IndustryFaqInputSchema] = Field(
        default_factory=list,
    )
    seo: IndustrySeoInputSchema = Field(
        default_factory=IndustrySeoInputSchema,
    )


class IndustryUpdateSchema(IndustryCreateSchema):
    change_summary: str = Field(
        default="",
        max_length=300,
    )


class IndustryScheduleSchema(Schema):
    scheduled_for: datetime


class IndustrySchema(Schema):
    id: UUID
    name: str
    slug: str
    short_description: str
    description: dict[str, Any]
    hero_title: str
    hero_description: str
    hero_image_id: UUID | None
    icon: str
    status: str
    published_at: datetime | None
    scheduled_for: datetime | None
    is_featured: bool
    is_active: bool
    is_publicly_available: bool
    sort_order: int
    challenges: list[Any]
    solutions: list[Any]
    benefits: list[Any]
    cta_title: str
    cta_text: str
    cta_label: str
    cta_url: str
    current_revision_number: int
    services: list[dict[str, Any]]
    faqs: list[dict[str, Any]]
    seo: dict[str, Any] | None
    revisions: list[dict[str, Any]]
    created_at: datetime
    updated_at: datetime

File: apps/dashboard/src/features/industries-catalog/schemas.ts
import { z } from "zod";

import { industryStatuses } from "./types";

const industryServiceLinkSchema = z.object({
  id: z.string(),
  service_id: z.string(),
  service_title: z.string(),
  description: z.string(),
  sort_order: z.number(),
  is_featured: z.boolean(),
});

const industryFaqSchema = z.object({
  id: z.string(),
  question: z.string(),
  answer: z.string(),
  sort_order: z.number(),
});

const industrySeoSchema = z.object({
  id: z.string(),
  meta_title: z.string(),
  meta_description: z.string(),
  canonical_url: z.string(),
  robots_index: z.boolean(),
  robots_follow: z.boolean(),
  open_graph_title: z.string(),
  open_graph_description: z.string(),
  open_graph_image_id: z.string().nullable(),
  twitter_title: z.string(),
  twitter_description: z.string(),
  structured_data: z.record(z.string(), z.unknown()),
});

const industryRevisionSchema = z.object({
  id: z.string(),
  revision_number: z.number(),
  snapshot: z.record(z.string(), z.unknown()),
  change_summary: z.string(),
  created_at: z.string(),
});

export const catalogIndustrySchema = z.object({
  id: z.string(),
  name: z.string(),
  slug: z.string(),
  short_description: z.string(),
  description: z.record(z.string(), z.unknown()),
  hero_title: z.string(),
  hero_description: z.string(),
  hero_image_id: z.string().nullable(),
  icon: z.string(),
  status: z.enum(industryStatuses),
  published_at: z.string().nullable(),
  scheduled_for: z.string().nullable(),
  is_featured: z.boolean(),
  is_active: z.boolean(),
  is_publicly_available: z.boolean(),
  sort_order: z.number(),
  challenges: z.array(z.unknown()),
  solutions: z.array(z.unknown()),
  benefits: z.array(z.unknown()),
  cta_title: z.string(),
  cta_text: z.string(),
  cta_label: z.string(),
  cta_url: z.string(),
  current_revision_number: z.number(),
  services: z.array(industryServiceLinkSchema),
  faqs: z.array(industryFaqSchema),
  seo: industrySeoSchema.nullable(),
  revisions: z.array(industryRevisionSchema),
  created_at: z.string(),
  updated_at: z.string(),
});

export const paginatedIndustriesSchema = z.object({
  items: z.array(catalogIndustrySchema),
  pagination: z.object({
    page: z.number(),
    page_size: z.number(),
    total_items: z.number(),
    total_pages: z.number(),
  }),
});

export const messageSchema = z.object({
  status: z.string(),
  message: z.string(),
});
